src/java.base/share/classes/java/lang/Class.java

```diff
@@ -261,7 +261,7 @@ public String toString() {
 
     /**
      * Returns a string describing this {@code Class}, including
-     * information about modifiers and type parameters.
+     * information about modifiers, {@linkplain #isSealed() sealing}, and type parameters.
```

Comment:
If Class.toGenericString is a useful API point for describing the blessed order of modifiers in the JDK
perhaps the existing classes that describe modifiers could/should cross reference this method.

Reply:
Given potential lossiness of source modifiers presentation by java.lang.reflect, I don't think this method is a good host for describing the canonical modifier order. Mentioning? Probably. Describing? No.

Separately, but also related: while it does not seem impossible, this PR does not implement `non-sealed`.

Reply:
I think the best place, or least-bad place, to discuss the "modifier" ordering of sealed/non-sealed would be an informative note on Modifier.toString(int) -- "The sealed/non-sealed Java language modifiers are not represented in the class file as access flags and thus not modeled by this class [java.lang.reflect.Modifier] .... sealed/non-sealed should be presented in the same location as final."

Since it doesn't seem possible to do so, I did not attempt to relay "non-sealed" information in this PR :-)

Reply:
Thanks @pavelrappo; I'll explore incorporating functionality like this into the PR, probably next week.

Reply:
> I think the best place, or least-bad place, to discuss the "modifier" ordering of sealed/non-sealed would be an informative note on Modifier.toString(int) -- "The sealed/non-sealed Java language modifiers are not represented in the class file as access flags and thus not modeled by this class [java.lang.reflect.Modifier] .... sealed/non-sealed should be presented in the same location as final."

Filed JDK-8322979: Add informative discussion to Modifier.

Reply:
> Thanks @pavelrappo; I'll explore incorporating functionality like this into the PR, probably next week.

Pushed an initial cut of non-sealing support; will add more test cases later.

Reply:
Thanks for including `non-sealed`, Joe.

Regarding your implementation: you are right that only the direct superclass and direct superinterfaces need to be explored for being `sealed`. No need to explore the full ancestry. So, recursion in my [initial proposal] was superfluous.

That said, I have a question on [JLS 8.1.1.2], which you copied as inline commentary:

> It is a compile-time error if a class is declared non-sealed but has neither a sealed direct superclass nor a sealed direct superinterface.
>
> Thus, a subclass of a non-sealed class cannot itself be declared non-sealed.

I wonder if it can be clarified to account for hierarchies where a `non-sealed` class or `non-sealed` interface has a `non-sealed` direct superinterface, or a `non-sealed` class has a `non-sealed` direct superclass.

For example:

    sealed interface A permits B { }

    sealed interface I permits X { }

    non-sealed interface B extends A { }

    non-sealed interface X extends B, I { }

and

    sealed class A permits B { }

    sealed interface I permits C { }

    non-sealed class B extends A { }

    non-sealed class C extends B implements I { }

(CC'ing @GavinBierman)

[initial proposal]: https://github.com/openjdk/jdk/pull/17239#discussion_r1440867004
[JLS 8.1.1.2]: https://docs.oracle.com/javase/specs/jls/se21/html/jls-8.html#jls-8.1.1.2